File: frontend/src/services/deck.ts
import { api } from "./api";

export interface Deck {
  id: number;
  title: string;
  description: string | null;
  user_id: number;
  is_public: boolean;
  tags?: string[] | null;
  created_at: string;
  updated_at: string;
}

export interface DeckCreate {
  title: string;
  description?: string | null;
  is_public?: boolean;
  tags?: string[] | null;
}

export interface DeckUpdate {
  title?: string;
  description?: string | null;
  is_public?: boolean;
  tags?: string[] | null;
}

export interface DeckListResponse {
  items: Deck[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface DeckListParams {
  page?: number;
  page_size?: number;
  is_public?: boolean;
  search?: string;
}

export interface DeckStatus {
  next_review: string | null;
  due_count: number;
  is_due: boolean;
  is_mastered: boolean;
}

export const deckService = {
  async list(params: DeckListParams = {}): Promise<DeckListResponse> {
    const response = await api.get<DeckListResponse>("/decks", { params });
    return response.data;
  },

  async getPublic(params: DeckListParams = {}): Promise<DeckListResponse> {
    const response = await api.get<DeckListResponse>("/decks/public", { params });
    return response.data;
  },

  async getById(id: number): Promise<Deck> {
    const response = await api.get<Deck>(`/decks/${id}`);
    return response.data;
  },

  async create(data: DeckCreate): Promise<Deck> {
    const response = await api.post<Deck>("/decks", data);
    return response.data;
  },

  async update(id: number, data: DeckUpdate): Promise<Deck> {
    const response = await api.patch<Deck>(`/decks/${id}`, data);
    return response.data;
  },

  async delete(id: number): Promise<void> {
    await api.delete(`/decks/${id}`);
  },

  async getStatus(deckId: number): Promise<DeckStatus> {
    const response = await api.get<DeckStatus>(`/decks/${deckId}/status`);
    return response.data;
  },
};

File: backend/app/schemas/deck.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field, field_validator


class DeckBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=200, description="Deck title")
    description: str | None = Field(None, max_length=1000, description="Deck description")
    is_public: bool = Field(default=False, description="Whether the deck is publicly visible")
    tags: list[str] | None = Field(default=None, description="Optional list of tags for the deck")

    @field_validator("tags", mode="before")
    @classmethod
    def parse_tags(cls, v: str | list[str] | None) -> list[str] | None:
        if v is None:
            return None
        if isinstance(v, list):
            return v
        if isinstance(v, str):
            return [t.strip() for t in v.split(",") if t.strip()]
        return None


class DeckCreate(DeckBase):
    pass


class DeckUpdate(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = Field(None, max_length=1000)
    is_public: bool | None = None
    tags: list[str] | None = None


class DeckResponse(DeckBase):
    id: int
    user_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DeckListResponse(BaseModel):
    items: list[DeckResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class DeckStatusResponse(BaseModel):
    next_review: Optional[datetime] = None
    due_count: int
    is_due: bool
    is_mastered: bool
